### DISCUSSION TITLE
Adding a reference to a remote model

### DISCUSSION THREAD

**Discussion by aparedes1204** [eclipse-emfcloud/emfcloud#208] [ModelServer]
Hi,

I'm working with the ModelServer REST API and I'm having some issues adding a reference to another model. I've been following the solution provided in another similar discussion, however it didn't work for me.

I have the following models:

**modelServerTestLibraryJSON.mlib**
```json
{
            "$type": "http://www.example.org/library#//Library",
            "$id": "_MecH4AZ4Ee2WyskjrqGP3w",
            "description": "JSON library description",
            "name": "JSON library",
            "users": {
                "$id": "_PK9YgAZ4Ee2WyskjrqGP3w",
                "users": [
                    {
                        "$type": "http://www.example.org/user#//User",
                        "$ref": "modelServerTestUsersJSON.muser#_F0qXgAZ4Ee2WyskjrqGP3w"
                    },
                    {
                        "$type": "http://www.example.org/user#//User",
                        "$ref": "modelServerTestUsersJSON.muser#_IYUAAAZ4Ee2WyskjrqGP3w"
                    }
                ]
            },
            "books": [
                {
                    "$id": "_Pf5twAZ4Ee2WyskjrqGP3w",
                    "name": "JSON book 1",
                    "description": "JSON book description 1",
                    "isbn": "1"
                },
                {
                    "$id": "_SCb78AZ4Ee2WyskjrqGP3w",
                    "name": "JSON book 2",
                    "description": "JSON book description 2",
                    "isbn": "2"
                }
            ]
}
```

**modelServerTestUsersJSON.muser**

```json
{
            "$type": "http://www.example.org/user#//Users",
            "$id": "_FRTLwAZ4Ee2WyskjrqGP3w",
            "users": [
                {
                    "$id": "_F0qXgAZ4Ee2WyskjrqGP3w",
                    "description": "JSON user description 1",
                    "identifier": 1,
                    "name": "JSON user 1"
                },
                {
                    "$id": "_IYUAAAZ4Ee2WyskjrqGP3w",
                    "description": "JSON user description 2",
                    "identifier": 2,
                    "name": "JSON user 2"
                },
                {
                    "$id": "_-t0FQAHEEe2eYq3VYGw6UA",
                    "description": "user3 desc",
                    "identifier": 3,
                    "name": "user3"
                },
                {
                    "$id": "_-t0sUAHEEe2eYq3VYGw6UA",
                    "description": "user4 desc",
                    "identifier": 4,
                    "name": "user4"
                },
                {
                    "$id": "_-t0sUQHEEe2eYq3VYGw6UA",
                    "description": "user5 desc",
                    "identifier": 5,
                    "name": "user5"
                }
            ]
}
```

Now, I'm tryng to add a new referenced user to the **modelServerTestLibraryJSON.milb** model with the following JSON Patch

```json
{
                "op": "add",
                "path": "/users/users/-",
                "value": {
                    "$type": "http://www.example.org/user#//User",
                    "$ref": "modelServerTestUsersJSON.muser#_-t0sUAHEEe2eYq3VYGw6UA"
                }
}
```

The request doesn't rise any errors, however this is the response:

```json
{
    "type": "success",
    "data": {
        "message": "Model 'modelServerTestLibraryJSON.mlib' successfully updated",
        "patch": [
            {
                "op": "add",
                "path": "/users/users/-",
                "value": {
                    "$type": "http://www.example.org/user#//User",
                    "$ref": "#//"
                }
            }
        ],
        "allPatches": [
            {
                "modelUri": "file:/path/to/model/modelServerTestLibraryJSON.mlib",
                "patch": [
                    {
                        "op": "add",
                        "path": "/users/users/-",
                        "value": {
                            "$type": "http://www.example.org/user#//User",
                            "$ref": "#//"
                        }
                    }
                ]
            }
        ]
    }
}
```

My guess is that the issue must be with the `$ref` field of the JSON Patch but I've try using different variations and examples that I found online and none worked.

I hope someone can help me with this, thanks in advance.

**Comment by ndoschek** [comment]
Hi @aparedes1204,
sorry for the late reply.
I'm afraid, we currently did not have a use case like that yet.
Could you maybe share your test models, so we can look into it in more detail?
Thank you in advance, BR Nina

**Reply by aparedes1204** [reply]
Hi, sure, the test models are modelServerTestUsersJSON.muser and modelServerTestLibraryJSON.mlib, those shown on the previous post. Or do you perhaps mean the *.ecore models I'll attach them.
[testModels.zip](https://github.com/eclipse-emfcloud/emfcloud/files/11488111/testModels.zip)

**Reply by ndoschek** [reply]
Thank you for the ecore models! I'll have a look into it!
Best, Nina

**Reply by ndoschek** [reply]
Hi @aparedes1204,
Thanks again for providing your ecore models!
I had a look at the scenario and debugged json patch command execution on the Model Server and discovered, that this case is unfortunately not covered in the implementation.
If you have a look at [the method `getAddReferenceCommand`](https://github.com/eclipse-emfcloud/emfcloud-modelserver/blob/13c848d37efe7008ab6929e8272a0536ab860973/bundles/org.eclipse.emfcloud.modelserver.common/src/org/eclipse/emfcloud/modelserver/common/patch/AbstractJsonPatchHelper.java#LL269C6-L269C6): it tries to determine the eObject to add via the passed `modelURI` (which in our case is the wrong one, it is the model URI of the library model). It can't resolve the passed reference properly to an object, hence the command is not executable and the user object is not added to the array of users.

If this is still relevant to you, would you mind open an issue in the [emfcloud-modelserver](https://github.com/eclipse-emfcloud/emfcloud-modelserver/issues)? TIA :sparkles: 
Also, if you would like to look into this we would gladly accept contributions :rocket: 

HTH and best regards,
Nina

**Comment by eneufeld** [comment]
Please reopen if issue still persists
